[SILOptimizer] Don't lower AssignInst in DI

Azoy opened the pull request:
Rather than lowering AssignInst in DI, mark the initialization kind in DI and instead lower them in the RawSILInstLowering pass.

I decided to do the bare minimal to accomplish this, but there are a couple of changes we could make depending on what other people think. `MandatoryOptUtils` is only used for `lowerAssignInstruction` now, so we could get rid of this file and move that function into `RawSILInstLowering` (which is the only file that uses that function now). In the bug report, Slava mentions the different init kinds which are `Unknown`, `Init`, `Reinit`, and `Assign`, so we could rename the enum cases in `PartialInitializationKind` to those respectably. The report also states that `Assign` init kind should lower to an assign store, but the existing lowering used a take load, init store, and destroyed the load so I decided not to touch it, but it might be that we can just change this now.

Also since this *should* be NFC, there shouldn't have to be any test updates, although I'm happy to add/adjust any tests that seem fit. 

cc: @slavapestov @gottesmm 

Solves: [SR-9466](https://bugs.swift.org/browse/SR-9466)

Azoy committed:
[SILOptimizer] Don't lower AssignInst in DI

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

Azoy commented:
@slavapestov is there any reason why the test wouldn't fail locally but would fail on CI? I've been running this test locally and I cannot reproduce this.

gottesmm commented:
@Azoy I would prefer if you did the store [assign] change. store [assign] will be lowered by the ownership model eliminator pass that runs before PMO to that pattern.

gottesmm commented:
Also, we should get rid of MandatoryOptUtils if we can.

Azoy commented:
@gottesmm are there any test recommendations that you have since I'm adding the store [assign] change?

Azoy committed:
Lower Assign to store [assign] & Remove MandatoryOptUtils

update debug_type

gottesmm reviewed (commented):
There are a few more things that are needed here:

1. We need to represent this flag in textual and serialized sil, not only in in-memory sil. I would just follow the example of the StoreOwnershipQualifier flag on Store. The printing/serializing/parsing should be copy/paste/modify. This will need printing/serializing/parsing tests.

2. Once you have that, the DI tests should fail since they shouldn't pattern match such assigns, unless we are missing code coverage. If we aren't missing code-coverage I would just update the DI tests directly for the new manner of printing assign inst. If we are missing code-coverage, lets go over that bridge when we get there = ).

3. I would add a test to lower raw sil insts one for each of the partial initialization kinds (except unknown). You want to show that an assign of each kind lowers appropriately. I think it would probably be ok to assert on assign unknown (I would be surprised if that is a valid outcome of DI).

Azoy commented on lib/SILOptimizer/Mandatory/RawSILInstLowering.cpp:
```diff
@@ -10,15 +10,81 @@
+
+  SILValue Src = Inst->getSrc();
+  SILLocation Loc = Inst->getLoc();
+  PartialInitializationKind initKind = Inst->getInitKind();
```
Throughout the whole file or just `lowerAssignInstruction`?

gottesmm replied:
Can you fix the inconsistency in this pass to lowerCamel? Then I will start the testing.

Azoy replied:
How does it look now? I only kept things like MUI, B, and BB uppercased, but I can lower case those as well if you want.

gottesmm replied:
That would be great.

Azoy commented:
@gottesmm how do you want to represent the initialization kinds? Do you want to do something like `[initialization]`, `[reinitialization]`, and `[assignment]`?

Azoy committed:
Textualize assign init kind

Rename [assign] to [reassign]

fix some tests

AssignOwnershipQualifier

formatting

moar formatting

Azoy commented:
@gottesmm I believe I got a lot of everything you said to add (if not everything). We can't assert on unknown because that's what is set when we parse an assign with no init kind. So in an only raw-sil-inst-lowering pass test we need to make sure that we give unknown the [assign] init kind (this mimics what was already in place here: https://github.com/apple/swift/blob/master/lib/SILOptimizer/Mandatory/RawSILInstLowering.cpp#L33). Would love if either you or @slavapestov took a look.

gottesmm commented:
@Azoy Maybe instead of using [assign], use [full]? Or maybe @jrose-apple has a better name? He is good at naming things usually.

jrose-apple commented:
`[replace]`, `[write]`? It's too bad `store` is something else, because otherwise I'd say `store [assign]` and `store [init]` would make the most sense.

jrose-apple commented:
I'm used to borrowing C++ terminology, where "assign" does mean there was something there before (cf. assignment operators). But the existence of `assign [init]` shows that we're not only using it that way today.

Azoy commented on lib/SILOptimizer/Mandatory/RawSILInstLowering.cpp:
```diff
@@ -30,11 +94,16 @@ static bool lowerRawSILOperations(SILFunction &Fn) {
+      // initialization.
+      //
+      // Unknown is considered unprocessed and is just Assign
+      // Init becomes a store [init]
```
I was trying to list each lowering based off init kind, but I can condense it if you want.

Azoy commented on include/swift/SIL/SILInstruction.h:
```diff
@@ -3592,6 +3592,22 @@ class EndUnpairedAccessInst
+  Unknown,
+
+  /// The box contains a fully-initialized value.
+  Assign,
```
Reassign sounds really nice.

Azoy commented on lib/SILOptimizer/Mandatory/RawSILInstLowering.cpp:
```diff
@@ -10,15 +10,79 @@
+  // Unknown initKind is considered unprocessed. Just lower it as
+  // Assign initialization
+  if (initKind == PartialInitializationKind::Unknown) {
+    initKind = PartialInitializationKind::Assign;
```
Yes, if we went through DI we should assert, but the `raw_sil_inst_lowering.sil` test doesn't go through DI and defaulted unknown init kinds to [reassign]. Here: https://github.com/apple/swift/blob/master/lib/SILOptimizer/Mandatory/RawSILInstLowering.cpp#L33
```cpp
      // Unprocessed assigns just lower into assignments, not initializations.
      if (auto *AI = dyn_cast<AssignInst>(Inst)) {
        SILBuilderWithScope B(AI);
        lowerAssignInstruction(B, AI,
                               PartialInitializationKind::IsNotInitialization);
        // Assign lowering may split the block. If it did,
        // reset our iteration range to the block after the insertion.
        if (B.getInsertionBB() != &BB)
          I = E;
        Changed = true;
        continue;
      }
```

gottesmm replied:
I know what the problem is. The previous behavior must have been that if it was unknown then it was a full mark uninitialize (and hence the code). This I think would only come up in situations where there was a memory location without a mark_uninitialize. IMO, this sounds like something where we will need to disambiguate these test cases in SILGen better (from looking at your code, SILGen only creates this instruction in one place).

So lets keep the old behavior, but can you add a nice comment on top that explains the situation and adds a TODO that says that we should fix SILGen and change this to an assert?

gottesmm commented:
I was thinking about this last night:

1. It might make sense to rename PartialInitializationKind to AssignOwnershipQualifier to match Store. Having the name PartialInitializationKind almost feels like the DI code leaking into the SIL namespace. The partial initialization aspect is pure DI. We are trying to describe how the Assign effects of the ownership of the underlying object stored into the memory.
2. When ever one changes serialization, one needs to increment the minor module format to prevent incompatibilities. That is missing in this commit. The place to change is here: https://github.com/apple/swift/blob/master/include/swift/Serialization/ModuleFormat.h#L55.

I think that once we resolve those two things, the commit LGTM.

Azoy commented:
@gottesmm picking this back up again, I think for the most part this looks ok, but I wanted to ask if there any SILVerifier changes that I should add in?

gottesmm commented:
Let me catch up with this thread real quick... That being said, can you fix up the git history slightly so that the rename commits are in the original relevant commits?

gottesmm commented:
I don't think you need to do any changes to the SILVerifier. I think the current set of AssignInst semantics should be fine.

Azoy commented:
@gottesmm @slavapestov Can either of you invoke CI for this please? Does this look ready?

jrose-apple commented:
CI without review

@swift-ci Please test

Azoy commented:
@gottesmm How does everything look now?

gottesmm commented:
Looks beautiful! Lets get this sucker in!

gottesmm commented:
@swift-ci smoke test and merge

gottesmm commented:
@swift-ci smoke test and merge
